### DISCUSSION TITLE
unit test case getting error on constant PIMCORE_PROJECT_ROOT - assumed 'PIMCORE_PROJECT_ROOT' #16361

### DISCUSSION THREAD

**Discussion by Thejaswini-Rao-U** [pimcore/pimcore#16421] [General]
**Pimcore version**
pimcore 6.6

**Steps to reproduce**
im trying to unit test by codeception, i have installed codception and tests/ folder created under my project directory. But when im calling the service function getting below error

**Actual Behavior**
Error attached

**[PHPUnit\Framework\Exception] Use of undefined constant PIMCORE_PROJECT_ROOT - assumed 'PIMCORE_PROJECT_ROOT' (this will throw an Error in a future version of PHP)

https://github.com/pimcore/pimcore/pull/1 src/GenericBundle/Service/PdfImportService.php:31**

File name for test is PdfImportServiceTest.php

public function testGetFilesArray()
{
$pdfService = new PdfImportService();
$this->assertFalse($pdfService->getFilesArray());
}

when im executing this test case, its calling constructor of actual service file and there we have defined constant path for the file, but this test case is not working. Actual function from the backend is working

Please let me know how to resolve it.

Thanks in advance

**Comment by Thejaswini-Rao-U** [comment]
Anyone please help me to solve tthis?

**Comment by kingjia90** [comment]
Do your tests have this part in the bootstrap?
https://github.com/pimcore/pimcore/blob/4f37f72d95e41a2c4720f4da340aa0f091a1a91a/tests/_bootstrap.php#L19-L31

**Comment by Thejaswini-Rao-U** [comment]
Hi
Thanks for the response.

i have added below code in _bootstrap.php inside tests/ path


use Pimcore\Tests\Util\Autoloader;

// define project root which will be used throughout the bootstrapping process
**define('PIMCORE_PROJECT_ROOT', realpath(__DIR__ . '/..'));

// set the used pimcore/symfony environment
putenv('PIMCORE_ENVIRONMENT=test');

require_once PIMCORE_PROJECT_ROOT . '/vendor/autoload.php';

\Pimcore\Bootstrap::setProjectRoot();
\Pimcore\Bootstrap::bootstrap();


// add the core pimcore test library to the autoloader - this could also be done in composer.json's autoload-dev section
// but is done here for demonstration purpose
require_once PIMCORE_PROJECT_ROOT . '/vendor/pimcore/pimcore/tests/_support/Util/Autoloader.php';

Autoloader::addNamespace('Pimcore\Tests', PIMCORE_PROJECT_ROOT . '/vendor/pimcore/pimcore/tests/_support');**

shall i add your code in bootstrap file?

and in my **codeception.dist.yml** code attached below

namespace: Tests
support_namespace: Support
actor_suffix: Tester
paths:
    tests: .
    output: ./_output
    data: ./Support/Data
    support: ./Support
    envs: ./_envs
settings:
    bootstrap: bootstrap.php
    colors: true
params:
    - env
extensions:
    enabled:
        - Codeception\Extension\RunFailed
        
 also i have added bootstrap.php file as below
 
 **define('PIMCORE_PROJECT_ROOT', realpath(__DIR__ . '/..'));
include PIMCORE_PROJECT_ROOT . '/vendor/autoload.php';

\Pimcore\Bootstrap::setProjectRoot(PIMCORE_PROJECT_ROOT);
\Pimcore\Bootstrap::bootstrap();**

can you please tell which file bootstrap.php or _bootstrap.php needs to points in the **codeception.dist.yml** file ?

**Reply by kingjia90** [reply]
We usually point to `_bootstrap.php`, see
https://github.com/pimcore/pimcore/blob/11.x/codeception.dist.yml#L4
https://github.com/pimcore/pimcore/blob/11.x/tests/_bootstrap.php

your `tests/bootstrap.php` lacks of the all the fallbacks mentioned in https://github.com/pimcore/pimcore/discussions/16421#discussioncomment-8045477 but you could just switch to `_bootstrap.php` and should also work

**Comment by Thejaswini-Rao-U** [comment]
@kingjia90 I have added code as you mentioned in the _bootstrap.php and pointed this file in the codeception.dist.yml file

still gives same error as i mentioned above **[PHPUnit\Framework\Exception] Use of undefined constant PIMCORE_PROJECT_ROOT - assumed 'PIMCORE_PROJECT_ROOT' (this will throw an Error in a future version of PHP)**

**Comment by Thejaswini-Rao-U** [comment]
Hi
Anyone please need help
